src/js/repeat.js

```diff
@@ -133,40 +133,23 @@ export default {
             return 0;
         }
 
-        let checkEl = el.parentElement.closest( '.or-repeat' );
-        const info = el.classList.contains( 'or-repeat-info' );
-        let count = info ? 1 : Number( el.querySelector( '.repeat-number' ).textContent );
-        const name = el.dataset.name || el.getAttribute( 'name' );
+        const isInfoElement = el.classList.contains( 'or-repeat-info' );
 
+        const toCount = isInfoElement ? `.or-repeat-info[data-name="${el.dataset.name}"]` : `.or-repeat[name="${el.getAttribute( 'name' )}"]`;
```

Comment:
The big win for me is centralizing all differences based on whether we're dealing with an info element or a real repeat instance. This way the algorithmic part is entirely shared and that reduces the cognitive load a lot for me.

I also really got tripped up by the `name` declaration before. The fact that it was picking up a different name based on whether or not the element was an info element was implicitly based on the presence of `el.dataset.name`. That felt like it added a lot of mental hoops to jump through.

Reply:
1. It wasn't entirely clear to me what a "repeat info" is supposed to mean. After some hunting, I found this in the docs:

    > Each repeat series has a sibling .or-repeat-info element that stores info that is relevant to that series.

    This still didn't clear much up for me, so I went and looked for these nodes as rendered. To validate what I think I understand, I think they provide two purposes?

    - Render a button bound to create new repeat instances
    - Specify a model reference (`data-name`) to the associated repeat structure, which is repeated when the button is clicked

2. It's not clear to me why that element would factor into a count. I would think this condition is only to determine the repeat's name, but the selector would be looking only for `.or-repeat` elements with the specified name, e.g.

    ```js
    const toCount = isInfoElement ? `.or-repeat[name="${el.dataset.name}"]` : `.or-repeat[name="${el.getAttribute( 'name' )}"]`;
    ```

    But it's entirely possible I'm misunderstanding what is being counted.

3. Maybe overly nitpicky, but I think it would be easier to follow the rest of this if the name specified that it's a selector. Names like `toCount` confuse me because I would expect them to specify some sort of iterable structure (set, list, queue, etc) of data to be counted.

Reply:
1. I also had to go hunting for that same doc and came to a similar conclusion. Would be good to get confirmation that this covers its full purpose and think about augmenting the existing docs a bit for posterity.

2. My understanding is that counting the `repeat-info` elements is useful for knowing how many repeat sequences of a particular type there are. Let's say you have a `butterfly` repeat nested in a `meadow` repeat. `meadow 1` has 10 butterflies, `meadow 2` has 4 butterflies. I believe there will be a total of 14 butterfly repeat instances and 2 corresponding `repeat-info`/buttons for adding butterflies/`butterfly` sequences. Basically it's the count of the immediate parent repeat, I think. I'm not sure, though, and it's one of the questions I've asked.

As far as I understand, what you wrote wouldn't make sense because a repeat element wouldn't have a `dataset.name`.

3. That's a good point. Does `toCountSelector` help?

Reply:
1. I agree, it would be good to add clarity to the docs. I think it would also be helpful to add a comment describing its purpose in `enketo-transformer`, as that's where I looked first.

2. I realized after posting my initial review that I may have been mistaken here. I think my confusion mostly arises from naming, and what this method is actually doing. It does make sense that one might want to get an "index" of the "info" button (for certain purposes), but it doesn't make sense (to me) that one would want to "count" it.

3. Definitely better, notwithstanding the general count/index naming confusion.

- - -

Regarding point 2, I tried to step through the test/implementation to understand what's going on, and I have to admit to only being more confused than when I started reviewing. It's not clear to me what `getIndex` is actually intended to "count", or what an "absolute index" (per JSDoc) represents.

I added some logging on calls from `add`, to hopefully gain some more insight. Here's what I see running the "three deep" test:

  <details>
    <summary>with current logic</summary>
<pre><code>
LOG: 'setup'
LOG: Object{name: '/data/outer', isInfoElement: true, result: 0}
LOG: Object{name: '/data/outer', isInfoElement: false, result: 0}
LOG: Object{name: '/data/outer/inner', isInfoElement: true, result: 0}
LOG: Object{name: '/data/outer/inner', isInfoElement: false, result: 0}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: true, result: 0}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: false, result: 0}
LOG: Object{name: '/data/outer/inner', isInfoElement: true, result: 0}
LOG: Object{name: '/data/outer/inner', isInfoElement: false, result: 0}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: true, result: 0}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: false, result: 0}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: true, result: 0}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: false, result: 0}
LOG: '/setup'
LOG: 'click third'
LOG: Object{name: '/data/outer/inner/third', isInfoElement: true, result: 0}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: false, result: 1}
LOG: '/click third'
LOG: 'click inner'
LOG: Object{name: '/data/outer/inner', isInfoElement: true, result: 0}
LOG: Object{name: '/data/outer/inner', isInfoElement: false, result: 1}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: true, result: 1}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: false, result: 2}
LOG: '/click inner'
LOG: 'click outer'
LOG: Object{name: '/data/outer', isInfoElement: true, result: 0}
LOG: Object{name: '/data/outer', isInfoElement: false, result: 1}
LOG: Object{name: '/data/outer/inner', isInfoElement: true, result: 1}
LOG: Object{name: '/data/outer/inner', isInfoElement: false, result: 2}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: true, result: 2}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: false, result: 3}
LOG: '/click outer'
</code></pre>
  </details>

  Here's what it looks like if I add two more clicks on `third`, once immediately and again after the `outer` click:

  <details>
    <summary>with added clicks</summary>
<pre><code>
LOG: 'click third'
LOG: Object{name: '/data/outer/inner/third', isInfoElement: true, result: 0}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: false, result: 1}
LOG: '/click third'
LOG: 'click third again'
LOG: Object{name: '/data/outer/inner/third', isInfoElement: true, result: 0}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: false, result: 2}
LOG: '/click third again'
LOG: 'click inner'
LOG: Object{name: '/data/outer/inner', isInfoElement: true, result: 0}
LOG: Object{name: '/data/outer/inner', isInfoElement: false, result: 1}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: true, result: 1}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: false, result: 3}
LOG: '/click inner'
LOG: 'click outer'
LOG: Object{name: '/data/outer', isInfoElement: true, result: 0}
LOG: Object{name: '/data/outer', isInfoElement: false, result: 1}
LOG: Object{name: '/data/outer/inner', isInfoElement: true, result: 1}
LOG: Object{name: '/data/outer/inner', isInfoElement: false, result: 2}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: true, result: 2}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: false, result: 4}
LOG: '/click outer'
LOG: 'click third after outer'
LOG: Object{name: '/data/outer/inner/third', isInfoElement: true, result: 0}
LOG: Object{name: '/data/outer/inner/third', isInfoElement: false, result: 3}
LOG: '/click third after outer'
</code></pre>
  </details>

These values are confusing, I've tried/given up half a dozen times to explain what I think they might mean. At this point I think I'm better off just sharing my confusion and hope for clarity in replies.

Reply:
I keep coming back to try to understand this and I think I have it, but if I do I think this may be a candidate for refactoring sooner rather than later. (Typing on phone now because I had to step away from desk, apologies if it’s a mess.)

I’d need to add a more complex test/model to validate all of this, but:

- It seems for “info” elements the index likely amounts to depth within a given clone action (or depth times however many repeats of the same path are specified? Are adjacent/redundant repeats of the same path supported?)
- For repeated model values it seems to correspond to position in the record top-to-bottom regardless of depth/structure. Which is to say I would think the final click in my modification pushed the inner/third and outer/inner/third indices to 5 and 6 respectively.

If I’m understanding this correctly (and I very well may still not!), I think this is two functions which just happen to produce mostly expected output by luck. And if that’s the case, I’d favor making this less DRY just for the opportunity to name and document the different semantics.

More than that, I think I would recommend starting to decouple state/DOM more aggressively than I would have at first glance. It doesn’t necessarily require adopting purpose built solutions right away (though it will probably add weight to those options if/when there’s bandwidth to discuss them). Just keeping a data structure of current state that’s represented in the DOM without referencing it, and clearly delineating how data flows.

It could also be a good opportunity to investigate framework-agnostic state management solutions (I have several in mind), but I’m hesitant to push for that just now for a bunch of reasons, unless there’s an appetite for it.

Ultimately the initial goals I have in mind, in order of hierarchy of needs:

- Clarify what this is actually expected to do, in discussion
- Prioritize naming/documenting improvements that reflect that clarification
- Identify and plan steps to allow reasoning with state separately from rendering
- Consider tools that might aid state/view in the current stack
- Plan longer term to evaluate whether tools built for these purposes might be introduced gradually to begin establishing a pattern for future iteration/maintenance

Reply:
Thanks! This is great. I agree that toCountSelector would be even clearer.

Reply:
The repeat-info was introduced when we started supporting zero repeats (but that's all I remember).

Reply:
Thanks, @MartijnR, that's helpful.

@eyelidlessness and I went over the logs he pulled again looking at the UI and we are feeling like we can now consolidate that with our understanding of `getIndex`. The confusion came from the last added click to add an instance of the third level of repeats. It adds a `third` instance to the first instance of `inner`, not to the last one.